cpp/src/arrow/compute/kernels/hash_aggregate_test.cc

```diff
@@ -174,81 +242,117 @@ Result<Datum> RunGroupBy(const BatchesWithSchema& input,
+      for (auto out_array : out_arrays) {
+        struct_fields.push_back(out_array[j]);
+      }
+      ARROW_ASSIGN_OR_RAISE(auto struct_array,
+                            StructArray::Make(struct_fields, output_schema->fields()));
+      struct_arrays.push_back(struct_array);
     }
-    sort_keys.emplace_back(static_cast<int>(i));
+    return ChunkedArray::Make(struct_arrays);
+  } else {
+    ArrayVector struct_fields(out_arrays.size());
+    for (size_t i = 0; i < out_arrays.size(); ++i) {
+      ARROW_ASSIGN_OR_RAISE(struct_fields[i], Concatenate(out_arrays[i]));
+    }
+    return StructArray::Make(std::move(struct_fields), output_schema->fields());
   }
-  std::shared_ptr<Schema> key_schema = schema(std::move(key_fields));
-  std::shared_ptr<Table> key_table = Table::Make(std::move(key_schema), key_columns);
-  SortOptions sort_options(std::move(sort_keys));
-  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Array> sort_indices,
-                        SortIndices(key_table, sort_options));
+}
 
-  ARROW_ASSIGN_OR_RAISE(
-      std::shared_ptr<Array> struct_arr,
-      StructArray::Make(std::move(out_arrays), output_schema->fields()));
+Result<Datum> RunGroupBy(const BatchesWithSchema& input,
+                         const std::vector<std::string>& key_names,
+                         const std::vector<std::string>& segment_key_names,
+                         const std::vector<Aggregate>& aggregates, bool use_threads,
+                         bool segmented = false, bool naive = false) {
+  if (segment_key_names.size() > 0) {
+    ARROW_ASSIGN_OR_RAISE(auto thread_pool, arrow::internal::ThreadPool::Make(1));
+    ExecContext seq_ctx(default_memory_pool(), thread_pool.get());
+    return RunGroupBy(input, key_names, segment_key_names, aggregates, &seq_ctx,
+                      use_threads, segmented, naive);
+  } else {
+    return RunGroupBy(input, key_names, segment_key_names, aggregates,
+                      threaded_exec_context(), use_threads, segmented, naive);
+  }
+}
 
-  return Take(struct_arr, sort_indices);
+Result<Datum> RunGroupBy(const BatchesWithSchema& input,
+                         const std::vector<std::string>& key_names,
+                         const std::vector<Aggregate>& aggregates, bool use_threads,
+                         bool segmented = false, bool naive = false) {
+  return RunGroupBy(input, key_names, {}, aggregates, use_threads, segmented);
 }
 
 /// Simpler overload where you can give the columns as datums
 Result<Datum> RunGroupBy(const std::vector<Datum>& arguments,
                          const std::vector<Datum>& keys,
-                         const std::vector<Aggregate>& aggregates,
-                         bool use_threads = false) {
+                         const std::vector<Datum>& segment_keys,
+                         const std::vector<Aggregate>& aggregates, bool use_threads,
+                         bool segmented = false, bool naive = false) {
```

Comment:
I wanted to test the case of segmented aggregation with empty segment-keys. We could decide to remove this.

Reply:
Agree that `segmented` seems redundant

Reply:
What does "naive" do here?

Reply:
`naive` refers to `NaiveGroupBy`. When `false`, the expected group-by output is [generally sorted](https://github.com/apache/arrow/pull/34311/files#diff-ec1e8907dfc3e636abecd3307879f331421e4f28f75e504f147826cebf04d7e0R167-R172).

Reply:
Sorry I am still confused - what is "naive group by"?

Reply:
> I wanted to test the case of segmented aggregation with empty segment-keys. We could decide to remove this.

What do you mean by "segmented aggregation with empty segment-keys"? Isn't this what happens when segment-keys is empty? (The original hash aggregate)

Reply:
> Sorry I am still confused - what is "naive group by"?

`NaiveGroupBy` is [a tester function](https://github.com/apache/arrow/blob/d5b3b4737838774db658d3c488fcd3e72bc13f7e/cpp/src/arrow/compute/kernels/hash_aggregate_test.cc#L75-L136) that computes the expected aggregation result (in a naive/simple/non-optimized way).

> What do you mean by "segmented aggregation with empty segment-keys"? Isn't this what happens when segment-keys is empty? (The original hash aggregate)

Let's take a step back. When `segment_keys` is non-empty the `segmented` flag is always `true`; otherwise (when empty), it may still be set to `true`, and this is the case we're discussing here. In this case, the tester restructures (without changing the data of) the result of `RunGroupBy` from `std::vector<ExecBatch> output_batches` to `std::vector<ArrayVector> out_arrays`, which have the structure typical of the case of a non-empty `segment_keys` (with multiple arrays per column, one array per segment) but only one array per column (because, technically, there is only one segment in this case). What all this boils down to is a test focusing on the structure, rather than the data, of the result.

Reply:
> therwise (when empty), it may still be set to true, and this is the case we're discussing here.

This will never happen in real scenario, right? If that is the case, I think we probably don't need to test that case. If you really want to test this then we should make it very clear in the test comment because this is quite confusing IMO.

Reply:
> NaiveGroupBy is [a tester function](https://github.com/apache/arrow/blob/d5b3b4737838774db658d3c488fcd3e72bc13f7e/cpp/src/arrow/compute/kernels/hash_aggregate_test.cc#L75-L136) that computes the expected aggregation result (in a naive/simple/non-optimized way).

I am still quite confused - what is this testing with `naive=true` vs `naive=false`?

Reply:
> This will never happen in real scenario, right?

The tester is trying to catch possible future errors, not just current ones. This scenario doesn't happen with the current implementation of the main code, but it could happen in a future implementation. I'll add documentation of the flag.

Reply:
> I am still quite confused - what is this testing with naive=true vs naive=false?

IIRC, we discussed this elsewhere in this PR. The `naive` flag means that the output is expected to be  that of like `NaiveGroupBy`, which in particular doesn't require sorting.

> Are we test a tester function here?

No. The reason for the `naive` flag is that the expected output of some test-cases (most, if not all, are pre-PR) is naive and of some others it is not. The common `RunGroupBy` function deals with both kinds of expected output. I'll add documentation for this.